### PR TITLE
OCPBUGS-31384: UPSTREAM: <carry>: allow type mutation for specific secrets

### PR DESCRIPTION
**Pull request by p0lyn0mial**
This PR relaxes the restrictions on allowed type mutation transitions.

initially, the check was stricter.
however, due to the platform's long history (spanning several years)
and the complexity of ensuring that resources were consistently created with only one type,
it is now permissible for (SecretTypeTLS, core.SecretTypeOpaque) type to transition to "kubernetes.io/tls".

additionally, it should be noted that default values might also be applied in some cases.
(https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/apis/core/v1/defaults.go#L280-L284)

### CODE REVIEW

File: pkg/apis/core/validation/validation_patch.go [hunk 1]
```diff
@@ -44,8 +44,14 @@ var (
...
-	// we allow "SecretTypeTLS" -> "kubernetes.io/tls" only
-	if oldSecret.Type == "SecretTypeTLS" && newSecret.Type == core.SecretTypeTLS {
+	// initially, this check was stricter.
+	// however, due to the platform's long history (spanning several years)
+	// and the complexity of ensuring that resources were consistently created with only one type,
+	// it is now permissible for any type to transition to "kubernetes.io/tls".
+	//
+	// additionally, it should be noted that default values might also be applied in some cases.
+	// (https://github.com/openshift/kubernetes/blob/258f1d5fb6491ba65fd8201c827e179432430627/pkg/apis/core/v1/defaults.go#L280-L284)
+	if oldSecret.Type != core.SecretTypeTLS && newSecret.Type == core.SecretTypeTLS {
```
Comment (tkashem) [inline pkg/apis/core/validation/validation_patch.go:54]: nit: do we care about the type of the old secret object - `if newSecret.Type == core.SecretTypeTLS {` ?

File: pkg/apis/core/validation/validation_patch_test.go [hunk 2]
```diff
@@ -52,45 +53,46 @@ func TestOpenShiftValidateSecretUpdate(t *testing.T) {
...
+				}
+
+				// exercise an invalid type mutation, we expect validation failure
+				errExpected = invalidTypeErrFn(secretType)
+				oldSecret, newSecret = newSecretFn(ns, name, "AnyType"), newSecretFn(ns, name, secretType)
+				if errGot := ValidateSecretUpdate(newSecret, oldSecret); !cmp.Equal(errExpected, errGot) {
+					t.Errorf("expected error: %v, diff: %s", errExpected, cmp.Diff(errExpected, errGot))
+				}
+
+				// verify that kbernetes.io/tls validation ar enforced
```
Comment (tkashem) [inline pkg/apis/core/validation/validation_patch_test.go:86]: typo: `are`

File: pkg/apis/core/validation/validation_patch.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go:54]: i think so, initially we didn't want to allow changes from `oldSecret.Type = core.SecretTypeTLS`

Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go:54]: now I think that maybe we should only allow to transition from “”, Opaque and SecretTypeTLS, especially that https://github.com/kubernetes/kubernetes/blob/8628c3c4da6746b1dc967cc520b189a04ebd78d1/pkg/apis/core/validation/validation.go#L6393

**Comment by tkashem** [comment]
/remove-label backports/unvalidated-commits
/label backports/validated-commits
/lgtm
/approve

/hold
(cancel if you want to merge it)

/cc @soltysh

File: pkg/apis/core/validation/validation_patch.go [hunk 3]
```diff
@@ -44,11 +44,27 @@ var (
...
 		if _, ok := whitelist[oldSecret.Namespace]; ok {
 			return true
 		}
 	}
 	return false
 }
+
+func isOldSecretTypeMutationAllowed(oldSecret *core.Secret) bool {
+	// "", core.SecretTypeOpaque seems safe because
+	//   https://github.com/kubernetes/kubernetes/blob/8628c3c4da6746b1dc967cc520b189a04ebd78d1/pkg/apis/core/validation/validation.go#L6393
```
Comment (benluddy) [inline pkg/apis/core/validation/validation_patch.go:64]: We should never see `""` in an the old secret of an update, right?

Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go:64]: Just playing it safe here, considering that these secrets could be created and then never changed, I don't think there is a way to ensure that resources were created only with one type.

Comment (benluddy) [reply pkg/apis/core/validation/validation_patch.go:64]: But if a client serializes a Secret that omits the type field or encodes it as `""`, its type would have been defaulted to Opaque. How would someone persist a Secret with type ""?

Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go:64]: isn't the default value applied on the server-side?

Comment (benluddy) [reply pkg/apis/core/validation/validation_patch.go:64]: Exactly. How would a Secret with type "" be persisted?

Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go:64]: Due to an incorrect patch applied to our server :) ?
Would you be okay to allow `core.SecretTypeOpaque` and `SecretTypeTLS` ?

Comment (p0lyn0mial) [reply pkg/apis/core/validation/validation_patch.go]: OK, thanks for your feedback, pushed, PTAL.

**Comment by p0lyn0mial** [comment]
/payload 4.16 ci blocking

sorry, @vrutkovs wanted to test blocking jobs

/hold

@tkashem please remove the hold once the payload job is green, thanks.

**Comment by vrutkovs** [comment]
/hold cancel

No regressions found in payload tests

**Pull request merged**